File: Kintsugi.Web/Features/Items/Item.cs
namespace Kintsugi.Web.Features.Items;

public class Item
{
    public int Id { get; set; }
    public string Title { get; private set; }
    public bool IsComplete { get; private set; }
    public DateTime CreatedAt { get; private set; }

    private Item()
    {
        Title = string.Empty;
    }

    public Item(string title) : this()
    {
        SetTitle(title);
        CreatedAt = DateTime.UtcNow;
    }

    public void Toggle()
    {
        IsComplete = !IsComplete;
    }

    public void Rename(string title)
    {
        SetTitle(title);
    }

    private void SetTitle(string title)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(title, nameof(title));
        Title = title;
    }
}

File: typescript/src/features/items/Item.ts
export class Item {
  constructor(
    public readonly id: number,
    public title: string,
    public isComplete: boolean,
    public readonly createdAt: Date,
  ) {}

  static create(title: string): Item {
    if (!title || title.trim().length === 0) {
      throw new Error("Title cannot be empty");
    }
    return new Item(0, title.trim(), false, new Date());
  }

  toggle(): void {
    this.isComplete = !this.isComplete;
  }

  rename(title: string): void {
    if (!title || title.trim().length === 0) {
      throw new Error("Title cannot be empty");
    }
    this.title = title.trim();
  }
}

export type ItemRow = {
  id: number;
  title: string;
  is_complete: number;
  created_at: string;
};
